src/goto-instrument/code_contracts.cpp

```diff
@@ -319,133 +316,109 @@ const symbolt &code_contractst::new_tmp_symbol(
+      freely_assignable_symbols.end())
   {
     return;
   }
-  exprt alias_expr = create_alias_expression(assigns, lhs, assigns_references);
 
   goto_programt alias_assertion;
   alias_assertion.add(goto_programt::make_assertion(
-    alias_expr, instruction_iterator->source_location));
+    assigns_clause.alias_expression(lhs),
+    instruction_iterator->source_location));
+
+  int lines_to_iterate = alias_assertion.instructions.size();
   program.insert_before_swap(instruction_iterator, alias_assertion);
-  ++instruction_iterator;
+  std::advance(instruction_iterator, lines_to_iterate);
 }
 
 void code_contractst::instrument_call_statement(
   goto_programt::instructionst::iterator &instruction_iterator,
   goto_programt &program,
   exprt &assigns,
-  std::vector<exprt> &aliasable_references,
-  std::set<exprt> &freely_assignable_exprs)
+  const irep_idt &function_id,
+  std::set<irep_idt> &freely_assignable_symbols,
+  assigns_clauset &assigns_clause)
 {
   INVARIANT(
     instruction_iterator->is_function_call(),
     "The first argument of instrument_call_statement should always be "
     "a function call");
+
   code_function_callt call = instruction_iterator->get_function_call();
   const irep_idt &called_name =
     to_symbol_expr(call.function()).get_identifier();
 
-  // Malloc allocates memory which is not part of the caller's memory
-  // frame, so we want to capture the newly-allocated memory and
-  // treat it as assignable.
   if(called_name == "malloc")
   {
-    aliasable_references.push_back(call.lhs());
-
-    // Make the variable, where result of malloc is stored, freely assignable.
     goto_programt::instructionst::iterator local_instruction_iterator =
       instruction_iterator;
+    // Malloc statments return a void pointer, which is then cast and assigned
+    // to a result variable. We iterate one line forward to grab the result of
+    // the malloc once it is cast.
     local_instruction_iterator++;
-    if(
-      local_instruction_iterator->is_assign() &&
-      local_instruction_iterator->get_assign().lhs().is_not_nil())
+    if(local_instruction_iterator->is_assign())
     {
-      freely_assignable_exprs.insert(
-        local_instruction_iterator->get_assign().lhs());
+      const exprt &rhs = local_instruction_iterator->get_assign().rhs();
+      if(rhs.id() == ID_typecast)
+      {
+        typet cast_type = rhs.type();
+
+        // Make freshly allocated memory assignable, if we can determine its type.
+        assigns_clause_targett *new_target =
+          assigns_clause.add_pointer_target(rhs);
+        goto_programt &pointer_capture = new_target->get_init_block();
+
+        int lines_to_iterate = pointer_capture.instructions.size();
+        program.insert_before_swap(local_instruction_iterator, pointer_capture);
+        std::advance(instruction_iterator, lines_to_iterate + 1);
+      }
+      else
+      {
+        log.error() << "Malloc is called but the result is not cast. "
```

Comment:
Does this also make sense for us to eliminate the if/else altogether, and just have an invariant guarding `rhs.id()`?